subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/IdeaProject.java

```diff
@@ -335,16 +325,10 @@ public void mergeXmlProject(Project xmlProject) {
     }
 
     private void configureModulePaths(Project xmlProject) {
-        ProjectComponentIdentifier thisProjectId = projectPathRegistry.getProjectComponentIdentifier(((ProjectInternal) project).getIdentityPath());
-        for (Path projectPath : projectPathRegistry.getAllProjectPaths()) {
-            ProjectComponentIdentifier otherProjectId = projectPathRegistry.getProjectComponentIdentifier(projectPath);
-            if (thisProjectId.getBuild().equals(otherProjectId.getBuild())) {
-                // IDEA Module for project in current build: handled via `modules` model elements.
-                continue;
-            }
-            LocalComponentArtifactMetadata imlArtifact = localComponentRegistry.findAdditionalArtifact(otherProjectId, "iml");
-            if (imlArtifact != null) {
-                xmlProject.addModulePath(imlArtifact.getFile());
+        Iterator<IdePlugin> it = project.getPlugins().withType(IdePlugin.class).iterator();
```

Comment:
This logic should live in the plugin. The plugin should create a lazy list of some kind and attach it to this model object, rather than having the model reach across and try to figure out which plugin created it. This removes some complexity.

Reply:
I was trying to move the logic out of there, but I kept running into backward compatibility. If we diverge too much from what we have now, we may break a lot of people that may be using the model in some untested ways. We may be alright with such breakage. I will look at the code again to see if I can remove this.

Reply:
The thing to do here would be to make `Project.modulePaths` as lazily configured property. That is, add a new property of type `SetProperty<Path>` and have the old methods delegate to this. Then, the IDEA plugin can add a `Provider<Path>` to this that provides the convention.

We'd do something similar on the Eclipse and Xcode (and Visual studio) models as well.

Oh, and we'd have to add `SetProperty` too.

I'd get the Xcode project generation happening correctly first in this PR, then get the Xcode workspace generation happening correctly in the next PR and then finally roll this out to the IDEA and Eclipse and Visual Studio plugins in a later PR (or multiple PRs).